pageserver/src/tenant/remote_timeline_client/download.rs

```diff
@@ -252,6 +252,104 @@ pub(super) async fn download_index_part(
+        }
+        Err(e) => {
+            if matches!(e, DownloadError::NotFound) {
+                tracing::debug!("No index_part found from previous generation {previous_gen:?}, falling back to listing");
+            } else {
+                return Err(e);
+            }
+        }
+    };
+
+    /// Given the key of an index, parse out the generation part of the name
+    fn parse_generation(path: RemotePath) -> Option<Generation> {
+        let file_name = match path.get_path().file_name() {
+            Some(f) => f,
+            None => {
+                // Unexpected: we should be seeing index_part.json paths only
+                tracing::warn!("Malformed index key {0}", path);
+                return None;
+            }
+        };
+
+        let file_name_str = match file_name.to_str() {
+            Some(s) => s,
+            None => {
+                tracing::warn!("Malformed index key {0}", path);
+                return None;
+            }
+        };
+
+        match file_name_str.split_once('-') {
+            Some((_, gen_suffix)) => u32::from_str_radix(gen_suffix, 16)
+                .map(Generation::new)
+                .ok(),
+            None => None,
+        }
+    }
+
+    // Fallback: we did not find an index_part.json from the previous generation, so
+    // we will list all the index_part objects and pick the most recent.
+    let index_prefix = remote_index_path(tenant_id, timeline_id, Generation::none());
+    let indices = backoff::retry(
+        || async { storage.list_files(Some(&index_prefix)).await },
+        |_| false,
+        FAILED_DOWNLOAD_WARN_THRESHOLD,
+        FAILED_REMOTE_OP_RETRIES,
+        "listing index_part files",
+        // TODO: use a cancellation token (https://github.com/neondatabase/neon/issues/5066)
+        backoff::Cancel::new(CancellationToken::new(), || -> anyhow::Error {
+            unreachable!()
+        }),
+    )
+    .await
+    .map_err(DownloadError::Other)?;
+
+    let mut generations: Vec<_> = indices
+        .into_iter()
+        .filter_map(parse_generation)
+        .filter(|g| g <= &my_generation)
+        .collect();
+
+    generations.sort();
+    match generations.last() {
+        Some(g) => {
+            tracing::debug!(
+                "Found index_part in generation {g:?} (my generation {my_generation:?})"
+            );
+            do_download_index_part(storage, tenant_id, timeline_id, *g).await
+        }
+        None => {
+            // This is not an error: the timeline may be newly created, or we may be
+            // upgrading and have no historical index_part with a generation suffix.
+            // Fall back to trying to load the un-suffixed index_part.json.
+            tracing::info!(
+                "No index_part.json-* found when loading {tenant_id}/{timeline_id} in generation {my_generation:?}"
+            );
+            do_download_index_part(storage, tenant_id, timeline_id, Generation::none()).await
```

Comment:
"the timeline may be newly created"

That doesn't make sense to me. In the generation world, timeline create will also use a generation, no?

Reply:
I was thinking of a situation where tenant/timeline creation had been interrupted and the control plane asks us to attach a timeline that we have local state for but nothing in remote storage.  But now I'm leaning toward thinking we should just assume that creation should always have uploaded an index_part before succeeding (and may be it already does).

Updated this comment to reflect the fact that this code path is only needed for migrations, where we find nothing in the index_part.json prefix, but the actual `index_part.json` with nothing after it may exist.